### ISSUE TITLE
Trouble with basic install

### ISSUE THREAD

**Issue by roygold** [reactjs/react-rails#184]
Hello, I'm struggling with a basic react-rails install in a simple/vanilla app.

Short version:

After an attempted install, Loading any page results in the error:

```
"couldn't find file 'react'
  (in .../app/assets/javascripts/application.js:16)
```
coming from the second line of the extracted source below (from app/views/layouts/application.html.erb) 
```
  <%= stylesheet_link_tag    'application', media: 'all', 'data-turbolinks-track' => true %>
  <%= javascript_include_tag 'application', 'data-turbolinks-track' => true %>

  <%= csrf_meta_tags %>
</head>
```
Also, if I try modifying config/environments/development.rb by adding 
```
config.react.variant = :development
```
then as soon as I start the rails server I get an error (personal paths changed to ... below)

```
$ bin/rails s -b 0.0.0.0
=> Booting WEBrick
=> Rails 4.2.0 application starting in development on http://0.0.0.0:3000
=> Run `rails server -h` for more startup options
=> Ctrl-C to shutdown server
Exiting
.../.rvm/gems/ruby-2.2.0/gems/railties-4.2.0/lib/rails/railtie/configuration.rb:95:in `method_missing': undefined method `react' for #<Rails::Application::Configuration:0x00000001a79d50> (NoMethodError)
	from .../config/environments/development.rb:42:in `block in <top (required)>'
	from .../.rvm/gems/ruby-2.2.0/gems/railties-4.2.0/lib/rails/railtie.rb:210:in `instance_eval'
	from .../.rvm/gems/ruby-2.2.0/gems/railties-4.2.0/lib/rails/railtie.rb:210:in `configure'
	from .../config/environments/development.rb:1:in `<top (required)>'
	from .../.rvm/gems/ruby-2.2.0/gems/activesupport-4.2.0/lib/active_support/dependencies.rb:274:in `require'
...
```

I have hit this with both 1.0.0.pre and 0.12.0, with no noticeable change. Here are the instructions I believe I've followed: https://github.com/reactjs/react-rails/blob/master/README.md

Here's some info about what I'm running:

```
$ ruby -v
ruby 2.2.0p0 (2014-12-25 revision 49005) [x86_64-linux]
$ rails -v
Rails 4.2.0
```

This line is at the bottom of my Gemfile:
```
gem 'react-rails', '~> 1.0.0.pre', github: 'reactjs/react-rails'
```
And I ran "bundle install", and here's the relevant parts of the output:
```
Using react-source 0.12.2
Using react-rails 1.0.0.pre from git://github.com/reactjs/react-rails.git (at master)
```
I also ran 
```
rails g react:install
```
Here's the relevant portion of app/assets/javascripts/application.js

```
...
// Read Sprockets README (https://github.com/sstephenson/sprockets#sprockets-directives) for details
// about supported directives.
//
//= require jquery
//= require jquery_ujs
//= require turbolinks
//= require react
//= require react_ujs
//= require components
//= require_tree .
```

The sample app works fine if I leave out the newly added "require" lines from application.js (or comment out javascript_include_tag line from app/views/layouts/application.html.erb).

Any suggestions would be very welcome; as I mentioned, I also tried 0.12 (based on the instructions here, and got the exact same results: https://github.com/reactjs/react-rails/blob/0.12/README.md )

Thanks!

**Comment by xionon** [comment]
I think spring has issues with gems installed directly from github

Try stopping and uninstalling spring and running your app again.

**Comment by roygold** [comment]
Thanks for the reply, but that doesn't seem to make a difference. I stopped the server, did a "gem uninstall spring," also tried "bin/spring binstub --remove --all" per instructions here: https://github.com/rails/spring. Removed spring from Gemfile, reran bundle install... spring is no longer showing up in "gem list". Then re-ran the rails server, but same results. Note I had the same problem when trying react-rails 0.12 as a local gem.

**Comment by rmosolgo** [comment]
Humm, I think react-rails expects to find that js `tmp/react-rails/react.js`

![image](https://cloud.githubusercontent.com/assets/2231765/6377658/902f1de6-bcda-11e4-94a1-7a5462004fda.png)

which is inserted by [an initializer](https://github.com/reactjs/react-rails/blob/master/lib/react/rails/railtie.rb#L33-L62) that initializer might be giving other troubles too (#148, #175). 

@roygold is that file present in your project?

**Comment by roygold** [comment]
There is no react-rails directory in tmp:

```
.../tmp$ ls
cache  pids  sessions  sockets
```

**Comment by rmosolgo** [comment]
huh, i wonder if it's running into a problem on that initializer or never running it at all :disappointed:

**Comment by jakubmal** [comment]
Yeah, I think we need to take a stab at the initializer again

**Comment by vipulnsward** [comment]
I think we can go ahead with trying https://github.com/reactjs/react-rails/pull/187

**Comment by roygold** [comment]
Hi all, I'm curious if you have any theories as to why I'm seeing this but others aren't (or don't seem to? I couldn't find any other open issues about this exact problem)? This happens every time for me on what feels like the most vanilla of basic install/example. Theories on what I'm doing differently that others aren't? If it matters, I'm running this on Ubuntu under VirtualBox on a recent MacBook Pro (2.8 GHz Intel Core i5). Thanks.

**Comment by rmosolgo** [comment]
I think the problem is that these lines aren't accomplishing their purpose:

https://github.com/reactjs/react-rails/blob/master/lib/react/rails/railtie.rb#L47-L50

They're supposed to copy React-related javascript files into `tmp/react-rails/`, but like you said, that folder isn't there. So ... how could it be failing?

- It never runs the initializer at all?
- It runs, but fails (and doesnt raise an error)?
- It runs and succeeds, but then somehow that directory is removed?

Regarding that first possibility, you could try checking out my branch in #187 and see if that improves your case.

Regarding the second possibility, there was some mention of file permissions issues getting in the way of the initializer (#175). Is there any chance of something along those lines in your setup?

As for the third possibility, I really don't expect that to happen!

**Comment by vipulnsward** [comment]
@roygold for now you can get away by defining config if any on `application.rb`

**Comment by vipulnsward** [comment]
btw @rmosolgo , the initializer is guaranteed to be run form rails, irrespective of not finding the after "sprockets" initializer hook.

@roygold can you create a sample vanilla app, as you said, which could demonstrate this issue.

**Comment by bensie** [comment]
Sitting here with @rmosolgo, I came up with a Rails 4.2.1 app that may demonstrate this issue. For me, the app works fine if I'm running from react-rails/master, but if I lock to a RubyGems version, I get the `couldn't find file 'react'` error you described.

https://github.com/bensie/react-rails-load-issue

Fails with https://github.com/bensie/react-rails-load-issue/blob/master/Gemfile#L26, but works if I change that line to `gem 'react-rails', '~> 1.0.0.pre', github: 'reactjs/react-rails'`.

Proposal: release master as 0.13.1.0 (aligns with current React version).

:heart: :heart: :heart:

**Comment by rmosolgo** [comment]
I took a try with @bensie's app:

```
rmosolgo ~/code $ git clone git@github.com:bensie/react-rails-load-issue.git
Cloning into 'react-rails-load-issue'...
rmosolgo ~/code $ cd react-rails-load-issue/
rmosolgo ~/code/react-rails-load-issue $ bundle install
Fetching gem metadata from https://rubygems.org/.........
rmosolgo ~/code/react-rails-load-issue $ bundle exec rails server
=> Booting WEBrick
=> Rails 4.2.1 application starting in development on http://localhost:3000
```

I got the same thing:

![image](https://cloud.githubusercontent.com/assets/2231765/6741648/9e44c066-ce46-11e4-8d41-41e0518f938c.png)

Then I switched to: 

```
gem 'react-rails', github: 'reactjs/react-rails'
```

And I got: 

![image](https://cloud.githubusercontent.com/assets/2231765/6741715/2ddc6ff8-ce47-11e4-96f9-2e7292feb90d.png)

(actually I had to add `<h1>It's working</h1>`)

I like @bensie's suggestion to release master as `0.13.1` -- clearly good stuff has happened in master and if we leave that out of these other versions, sadness happens.

**Comment by zpao** [comment]
I wouldn't release master as 0.13.1 (that's confusing when 0.13.0 and the rest of the 0.x.y family don't have any of the good stuff in master). Just ship 1.0 and forget about versioning with React versions. Just upgrade the dependency purposefully and bump the react-rails version appropriately when you do.

**Comment by bensie** [comment]
Cool that makes sense too. Should the README be updated to encourage specifying a custom "react-source" line in the Gemfile? That way it's clear that react-rails != react and it should work on any version greater than whatever is in the gemspec.

**Comment by zpao** [comment]
Leaving no upper cap isn't really safe either. For argument's sake lets say we were publishing master this whole time and you used react-rails which used `React.renderComponentToString`. That worked with the react-source specified. But then you specify a different react-source (like 0.13) and that API no longer exists so your app suddenly starts throwing 500 errors even though you didn't touch react-rails.

Unbounded dependencies are generally not a great idea. I would encourage locking to 0.x and if it turns out that 0.x+1 is also compatible, then ship and update that says so.

**Comment by bensie** [comment]
Totally agree. :+1:

**Comment by bezelga** [comment]
Hey guys, I got the same error when I try to run my Rails app in a docker container.

My react-rails version is:

`gem 'react-rails' , '~> 1.0'`

and Ruby 2.1.5

The strange thing is that running it locally on OSX it works normally.

Do you have an idea on why this is happening only when using docker?

**Comment by rmosolgo** [comment]
@bezelga react-rails writes some files into `tmp/react-rails`, can you see those files in the docker environment?

**Comment by bezelga** [comment]
@rmosolgo yes:

```bash 
$ ls tmp/react-rails/
JSXTransformer.js  react.js
```

**Comment by rmosolgo** [comment]
This is such a drag, check out #254 for a copy-less approach to app setup, I hope that will solve this problem!

**Comment by rmosolgo** [comment]
We've merged #254 which removed the copy-on-boot flow completely! Please update to master and give it a try.

**Comment by danillop** [comment]
Try to install nodejs, it works for me.
